// grunt-wordpress/Gruntfile.js
module.exports = function( grunt ) {
	//[[NEWCODE:hyunsook.park by updated]] - Add by being changed from the "lint" plugin to the "grunt-contrib-jshint" plugin.
	"use strict";

	//[[NEWCODE:hyunsook.park by updated]] - Add by being changed from the "lint" plugin to the "grunt-contrib-jshint" plugin.
	// grunt plugins
	grunt.loadNpmTasks( "grunt-contrib-jshint" );

	// Project configuration
	grunt.initConfig({
		//[[NEWCODE:hyunsook.park by updated]] - Add to migrate from Grunt v0.3.x to Grunt v0.4.x.
		pkg: grunt.file.readJSON( "package.json" ),
		jshint: {
			all: {
				options: {
					jshintrc: ".jshintrc"
				},
				files: {
					src: [ "Gruntfile.js", "tasks/**/*.js" ]
				}
			}
		}
		//[[OLDCODE:hyunsook.park by updated]] - Grunt v0.3.x
		//lint: {
		//	files: [ "grunt.js", "tasks/**/*.js" ]
		//},
		/*
		jshint: (function() {
			var rc = grunt.file.readJSON( ".jshintrc" ),
				settings = {
					options: rc,
					globals: {}
				};

			(rc.predef || []).forEach(function( prop ) {
				settings.globals[ prop ] = true;
			});
			delete rc.predef;

			return settings;
		})()
		*/
	});

	// Default grunt
	grunt.registerTask( "default", [ "jshint" ] );
};
