// grunt-jquery-content/Gruntfile.js
module.exports = function( grunt ) {
	"use strict";

	//[[NEWCODE:hyunsook.park by updated]] - Add by being changed from the "lint" plugin to the "grunt-contrib-jshint" plugin.
	// grunt plugins
	grunt.loadNpmTasks( "grunt-contrib-jshint" );

	// Project configuration.
	grunt.initConfig({
		//[[NEWCODE:hyunsook.park by updated]] - Add to migrate from Grunt v0.3.x to Grunt v0.4.x.
		pkg: grunt.file.readJSON( "package.json" ),
		jshint: {
			all: {
				options: {
					jshintrc: ".jshintrc"
				},
				files: {
					src: [ "Gruntfile.js", "tasks/**/*.js" ]
				}
			}
		},
		watch: {
			all: {
				files: [ "<%= jshint.all.files.src %>" ],
				tasks: [ "default" ]
			}
		}
		//[[OLDCODE:hyunsook.park by updated]] - Grunt v0.3.x
		//lint: {
		//	files: ['grunt.js', 'tasks/**/*.js' ]
		//},
		/*
		watch: {
			files: '<config:lint.files>',
			tasks: 'default'
		},
		jshint: {
			options: {
				curly: true,
				eqeqeq: true,
				immed: true,
				latedef: true,
				newcap: true,
				noarg: true,
				sub: true,
				undef: true,
				boss: true,
				eqnull: true,
				node: true,
				es5: true
			},
			globals: {}
		}
		*/
	});

	// Load local tasks.
	grunt.loadTasks( "tasks" );

	// Default task.
	grunt.registerTask( "default", [ "jshint" ] );
};
